### PR TITLE
Support nullable array type values as spark_udf return values

### LINKED ISSUE #8986: [BUG] Error when returning a null value for array types in spark_udf
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

- Client: 2.2.2

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu AWS
- **Python version**: 3.9
- **yarn version, if running the dev UI**:

### Describe the problem

When returning an array type in a model and using spark_udf, I get the following error message:

### Tracking information

_No response_

### Code to reproduce issue

Here is a minimal reproducible example:

```python
import mlflow
import numpy as np
import pandas as pd
import pyspark.sql.functions as F


class MyMLFLowModel(mlflow.pyfunc.PythonModel):
    def predict(self, context, model_input):
        model_input["B"] = list(np.random.rand(len(model_input), 100))
        # Insert null values
        model_input["B"] = [v if i % 2 == 1 else None for i, v in zip(model_input["A"], model_input["B"])]
        return model_input

with mlflow.start_run():
    model = MyMLFLowModel()
    model_info = mlflow.pyfunc.log_model("model", python_model=model)

model_udf = mlflow.pyfunc.spark_udf(spark, model_uri=model_info.model_uri, result_type="B array<double>")


test_df = spark.createDataFrame(pd.DataFrame(np.arange(10), columns=['A']))
pred_df = test_df.withColumn("prediction", model_udf(F.struct("*")))
pred_df.toPandas()
```

It works fine when creating my own UDF:

```python
from typing import Any, Union, Iterator, Tuple
import numpy as np
import pandas as pd
import pyspark.sql.functions as F

@F.pandas_udf("B array<double>")
def custom_udf(
    iterator: Iterator[Tuple[Union[pd.Series, pd.DataFrame], ...]]
) -> Iterator[pd.DataFrame]:
    for input_batch in iterator:
        if isinstance(input_batch, (pd.Series, pd.DataFrame)):
            # UDF is called with only one argument
            row_batch_args = (input_batch,)
        else:
            row_batch_args = input_batch

        if len(row_batch_args[0]) > 0:
            pred = model_loaded.predict(*row_batch_args)
            yield pred

test_df = spark.createDataFrame(pd.DataFrame(np.arange(10), columns=['A']))
pred_df = test_df.withColumn("prediction", custom_udf(F.struct("*"))).select("A", "prediction.*")
pred_df.toPandas()
```

### Stack trace

```python
Traceback (most recent call last):
  File "pyarrow/array.pxi", line 915, in pyarrow.lib.Array.from_pandas
  File "pyarrow/array.pxi", line 312, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 83, in pyarrow.lib._ndarray_to_array
  File "pyarrow/error.pxi", line 99, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Can only convert 1-dimensional array values
```

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by WeichenXu123**
## Related Issues/PRs

 Closes #8986

## What changes are proposed in this pull request?

Support nullable array type values as spark_udf return values

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/pyfunc/test_spark.py [hunk 1]
```diff
@@ -1014,3 +1014,25 @@ def predict(self, context, model_input, params=None):
...
+    with mlflow.start_run() as run:
+        mlflow.pyfunc.log_model(
+            "model",
+            python_model=TestModel(),
+        )
+        udf = mlflow.pyfunc.spark_udf(
+            spark,
+            f"runs:/{run.info.run_id}/model",
+            result_type="a array<long>",
+        )
```
Comment (harupy) [inline tests/pyfunc/test_spark.py:1035]:
```suggestion
        model_info = mlflow.pyfunc.log_model(
            "model",
            python_model=TestModel(),
        )
        udf = mlflow.pyfunc.spark_udf(
            spark,
            model_info.model_uri,
            result_type="a array<long>",
        )
```

**Comment by harupy** [comment]
@WeichenXu123 Can you run black before making commits?

**Comment by harupy** [comment]
@WeichenXu123 If pre-commit doesn't work, you can try https://akshay-jain.medium.com/pycharm-black-with-formatting-on-auto-save-4797972cf5de.

File: tests/pyfunc/test_spark.py [hunk 2]
```diff
@@ -1014,3 +1014,25 @@ def predict(self, context, model_input, params=None):
         )
         res = good_data.withColumn("res", udf("a")).select("res").toPandas()
         assert res["res"][0] == [("str", 0, 1, 0.0, 0.1, True)]
+
+
+def test_spark_udf_return_nullable_array_field(spark):
+    class TestModel(PythonModel):
+        def predict(self, context, model_input):
+            values = [np.array([1, 2])] * len(model_input)
+            values[-1] = None
```
Comment (harupy) [inline tests/pyfunc/test_spark.py:1023]:
```suggestion
```

File: mlflow/pyfunc/__init__.py [hunk 3]
```diff
@@ -936,9 +936,15 @@ def _convert_array_values(values, elem_type, array_dim, spark_primitive_type_to_
         )
 
     if array_dim == 1:
-        return [np.array(v, dtype=np_type) for v in values]
+        return [
+            np.array(v, dtype=np_type) if v is not None else None
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:940]:
```suggestion
            None if v is None else np.array(v, dtype=np_type) 
```

File: mlflow/pyfunc/__init__.py [hunk 4]
```diff
@@ -936,9 +936,15 @@ def _convert_array_values(values, elem_type, array_dim, spark_primitive_type_to_
...
+            for v in values
+        ]
     else:
-        return [list(np.array(v, dtype=np_type)) for v in values]
+        return [
+            list(np.array(v, dtype=np_type)) if v is not None else None
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:945]:
```suggestion
            None if v is None else list(np.array(v, dtype=np_type))
```

File: mlflow/pyfunc/__init__.py [hunk 5]
```diff
@@ -923,6 +923,11 @@ def _parse_spark_datatype(datatype: str):
     return udf(lambda x: x, returnType=datatype).returnType
 
 
+def _is_null_array(value):
+    # If provided value is None or NaN, regard it as a null array.
+    return value is None or isinstance(value, float) and np.isnan(value)
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:928]: Is `value` an array?

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:928]: yes in normal case.

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:928]: Note Nan is regarded as None, because https://github.com/mlflow/mlflow/issues/8986#issuecomment-1627960699

Comment (harupy) [reply mlflow/pyfunc/__init__.py:928]: why do we need `isinstance(value, float)`?

Comment (harupy) [reply mlflow/pyfunc/__init__.py:928]: I'd avoid using `array` because this function seems to work with a non-array value.

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:928]:
```
`np.isnan()` will raise error for values that is not number.  We need to add a check before calling `np.isnan()`
```

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:928]: what about the name `_is_null_element`?

Comment (harupy) [reply mlflow/pyfunc/__init__.py:928]: > `np.isnan()` will raise error for values that is not number. We need to add a check before calling `np.isnan()`

got it, this should be added as a comment

Comment (harupy) [reply mlflow/pyfunc/__init__.py:928]: can this function only be used to an element in an array?

Comment (harupy) [reply mlflow/pyfunc/__init__.py:928]: I'd call _is_none_or_nan

Comment (harupy) [reply mlflow/pyfunc/__init__.py:928]: @WeichenXu123 When naming a function, just describe what the function does. Don't need to assume how it's used.

File: tests/pyfunc/test_spark.py [hunk 6]
```diff
@@ -1014,3 +1014,24 @@ def predict(self, context, model_input, params=None):
         )
         res = good_data.withColumn("res", udf("a")).select("res").toPandas()
         assert res["res"][0] == [("str", 0, 1, 0.0, 0.1, True)]
+
+
+def test_spark_udf_return_nullable_array_field(spark):
+    class TestModel(PythonModel):
+        def predict(self, context, model_input):
+            values = [np.array([1, 2])] * (len(model_input) - 2) + [None, np.nan]
```
Comment (harupy) [inline tests/pyfunc/test_spark.py:1022]:
```suggestion
            values = [np.array([1, np.nan])] * (len(model_input) - 2) + [None, np.nan]
```

what would happen if we run this code?

Comment (WeichenXu123) [reply tests/pyfunc/test_spark.py]: works. , the np.nan inside the array is also converted to None.

File: mlflow/pyfunc/__init__.py [hunk 7]
```diff
@@ -923,6 +923,12 @@ def _parse_spark_datatype(datatype: str):
     return udf(lambda x: x, returnType=datatype).returnType
 
 
+def _is_none_or_nan(value):
+    # if value is not numeric type, `np.isnan(value)` raises error,
+    # so we add condition `isinstance(value, float)` before it.
+    return value is None or isinstance(value, float) and np.isnan(value)
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:929]: /ai simplify the comment in the following code.

Comment (harupy) [reply mlflow/pyfunc/__init__.py:929]: /ai simplify the comment in the following code. Please make sure to mention why `isinstance(value, float)` is needed.

**Review by harupy — APPROVED** [review]

**Pull request merged**